pkg/virt-launcher/virtwrap/util/libvirt_helper.go

```diff
@@ -251,71 +254,136 @@ func StartLibvirt(stopChan chan struct{}) {
-					}
-					time.Sleep(time.Second)
-				}
-				// #nosec No risk for path injection. logfile has a static basedir
-				file, err := os.Open(logfile)
-				if err != nil {
-					log.Log.Reason(err).Error("failed to catch virtlogd logs")
-					return
-				}
-				defer util.CloseIOAndCheckErr(file, nil)
+			scanner := bufio.NewScanner(file)
+			scanner.Buffer(make([]byte, 1024), 512*1024)
+			for scanner.Scan() {
+				log.LogQemuLogLine(log.Log, scanner.Text())
+			}
 
-				scanner := bufio.NewScanner(file)
-				scanner.Buffer(make([]byte, 1024), 512*1024)
-				for scanner.Scan() {
-					log.LogQemuLogLine(log.Log, scanner.Text())
-				}
+			if err := scanner.Err(); err != nil {
+				log.Log.Reason(err).Error("failed to read virtlogd logs")
+			}
+		}()
+
+		go func() {
+			defer close(exitChan)
+			_ = cmd.Wait()
+		}()
+
+		select {
+		case <-stopChan:
+			_ = cmd.Process.Kill()
+			return
+		case <-exitChan:
+			log.Log.Errorf("virtlogd exited, restarting")
+		}
 
-				if err := scanner.Err(); err != nil {
-					log.Log.Reason(err).Error("failed to read virtlogd logs")
-				}
-			}()
+		// this sleep is to avoid consumming all resources in the
+		// event of a virtlogd crash loop.
+		time.Sleep(time.Second)
+	}
+}
 
-			go func() {
-				defer close(exitChan)
-				cmd.Wait()
-			}()
+func startQEMUSeaBiosLogging(stopChan chan struct{}) {
+	const QEMUSeaBiosDebugPipeMode uint32 = 0666
+	const logLinePrefix = "[SeaBios]:"
+
+	err := syscall.Mkfifo(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed creating a pipe for sea bios debug logs: %s", logLinePrefix, err.Error()))
+		return
+	}
+
+	// Chmod is needed since umask is 0018. Therefore Mkfifo does not actually create a pipe with proper permissions.
+	err = syscall.Chmod(QEMUSeaBiosDebugPipe, QEMUSeaBiosDebugPipeMode)
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed executing chmod on pipe for sea bios debug logs: %s", logLinePrefix, err.Error()))
+		return
+	}
+
+	QEMUPipe, err := os.OpenFile(QEMUSeaBiosDebugPipe, os.O_RDONLY, 0666)
+	defer QEMUPipe.Close()
+
+	if err != nil {
+		log.Log.Reason(err).Error(fmt.Sprintf("%s failed to open %s with an error: %s", logLinePrefix, QEMUSeaBiosDebugPipe, err.Error()))
+	}
+
+	startTime := time.Now()
+	firstRun := true
+
+	var line []byte
+	reader := bufio.NewReader(QEMUPipe)
+
+	for {
+		for line, _, err = reader.ReadLine(); err == nil; line, _, err = reader.ReadLine() {
+			curTime := time.Now()
+			deltaTime := curTime.UnixNano() - startTime.UnixNano()
+
+			// Reset start time
+			if (time.Duration(deltaTime)*time.Nanosecond) > (60*time.Second) || firstRun {
+				firstRun = false
+
+				startTime = time.Now()
+				logLine := fmt.Sprintf("%s ==== Start time: %d ====", logLinePrefix, startTime.UnixNano())
+
+				log.LogQemuLogLine(log.Log, logLine)
+			}
+
+			logLine := fmt.Sprintf("%s [time passed from start: %d]: %s", logLinePrefix, deltaTime, string(line))
+
```

Comment:
Can you please explain what exactly the timing info you give here and why is it important?  
Why do we need delta from some starting point, why is that starting point being reset if the diff is > 60 sec? 
And in general why not just putting a timestamp?

Reply:
Basically I've tried to mimic the script's logic. I did that in the first place because Vladik said to do so by [following this explanation](https://www.seabios.org/Debugging#Timing_debug_messages).

@vladikr Can you answer that question?